runners/flink/src/test/java/org/apache/beam/runners/flink/streaming/DoFnOperatorTest.java

```diff
@@ -610,6 +620,252 @@ public void testSideInputs(boolean keyed) throws Exception {
+
+      TupleTag<String> outputTag = new TupleTag<>("main-output");
+
+      StringUtf8Coder keyCoder = StringUtf8Coder.of();
+
+      ImmutableMap<Integer, PCollectionView<?>> sideInputMapping =
+          ImmutableMap.<Integer, PCollectionView<?>>builder()
+              .put(1, view1)
+              .put(2, view2)
+              .build();
+
+      DoFnOperator<String, String> doFnOperator = new DoFnOperator<>(
+          new IdentityDoFn<>(),
+          "stepName",
+          coder,
+          outputTag,
+          Collections.emptyList(),
+          new DoFnOperator.MultiOutputOutputManagerFactory<>(outputTag, coder),
+          WindowingStrategy.globalDefault(),
+          sideInputMapping, /* side-input mapping */
+          ImmutableList.of(view1, view2), /* side inputs */
+          PipelineOptionsFactory.as(FlinkPipelineOptions.class),
+          keyCoder);
+
+      return new KeyedTwoInputStreamOperatorTestHarness<>(
+          doFnOperator,
+          new StringKeySelector(),
+          // we use a dummy key for the second input since it is considered to be broadcast
+          new DummyKeySelector(),
+          BasicTypeInfo.STRING_TYPE_INFO);
+    });
+  }
+
+  void pushbackDataCheckpointing(
+      TestHarnessFactory<
+          TwoInputStreamOperatorTestHarness<
+              WindowedValue<String>,
+              RawUnionValue,
+              WindowedValue<String>>> harnessFactory) throws Exception {
+
+    TwoInputStreamOperatorTestHarness<WindowedValue<String>, RawUnionValue, WindowedValue<String>>
+        testHarness = harnessFactory.create();
+
+    testHarness.open();
+
+    IntervalWindow firstWindow = new IntervalWindow(new Instant(0), new Instant(100));
+    IntervalWindow secondWindow = new IntervalWindow(new Instant(0), new Instant(500));
+
+    // push in main-input elements
+    WindowedValue<String> helloElement = valueInWindow("Hello", new Instant(0), firstWindow);
+    WindowedValue<String> worldElement = valueInWindow("World", new Instant(1000), firstWindow);
+    testHarness.processElement1(new StreamRecord<>(helloElement));
+    testHarness.processElement1(new StreamRecord<>(worldElement));
+
+    // snapshot state, throw away the operator, then restore and verify that we still match
+    // main-input elements to the side-inputs that we sent earlier
+    OperatorStateHandles snapshot = testHarness.snapshot(0, 0);
+
+    testHarness = harnessFactory.create();
+
+    testHarness.initializeState(snapshot);
+    testHarness.open();
+
+    // push in some side inputs for both windows
+    testHarness.processElement2(
+        new StreamRecord<>(
+            new RawUnionValue(
+                1,
+                valuesInWindow(
+                    ImmutableList.of(KV.of((Void) null, "hello"), KV.of((Void) null, "ciao")),
+                    new Instant(0),
+                    firstWindow))));
+    testHarness.processElement2(
```

Comment:
Just a clarification for myself: The use of 2 side inputs vs. one isn't relevant for the new tests, that's just a byproduct of using the existing setup?